js/cryptoLib/crypto.js

```diff
@@ -0,0 +1,249 @@
+/**
+ * AES GCM Stream
+ * This module exports encrypt and decrypt stream constructors which can be
+ * used to protect data with authenticated encryption.
+ *
+ * Helper methods are also provided to do things like generate secure keys
+ * and salt.
+ */
+'use strict';
+
+let stream = require('stream');
+let Transform = stream.Transform;
+let util = require('util');
+let crypto = require('crypto');
+
+let PBKDF2_PASS_LENGTH = 256;
+let PBKDF2_SALT_LENGTH = 32;
+let PBKDF2_ITERATIONS = 5000;
+let PBKDF2_DIGEST = 'sha256';
+let KEY_LENGTH = 32; // bytes
+let GCM_NONCE_LENGTH = 12; //bytes
+let GCM_MAC_LENGTH = 16; //bytes
+
+let keyEncoding = 'base64';
+
+/**
+ * Private helper method to validate a key passed into the Encrypt and Decrypt streams.
+ * Strings are converted it into a buffer, buffers are returned as they are.
+ * @param key
+ * @throws Missing, Encoding, or Length errors
+ * @returns Buffer
+ */
+let validateAndConvertKey = function(key) {
+    if (key && key instanceof Buffer && key.length === KEY_LENGTH) {
+        return key;
+    } else if (key && typeof key === 'string') {
+        // Because we don't have a reliable way to test string encoding, we assume that a string type
+        // key was created by the createEncodedKey method and that it is using the keyEncoding which
+        // has been set on the module, whether that's the default or something explicitly set by the
+        // user via the setKeyEncoding method.
+        let bufKey = new Buffer(key, keyEncoding);
+        if (bufKey.length !== KEY_LENGTH) {
+            let encodingErrorMessage = 'Provided key string is either of an unknown encoding (expected: ' +
+                keyEncoding + ') or the wrong length.';
+            throw new Error(encodingErrorMessage);
+        }
+        return bufKey;
+    } else {
+        let message = 'The key options property is required! Expected ' +
+            keyEncoding + ' encoded string or a buffer.';
+        throw new Error(message);
+    }
+};
+
+exports.encrypt = EncryptionStream;
+exports.decrypt = DecryptionStream;
+
+/**
+ * getEncoding
+ * Helper which returns the current encoding being used for keys
+ * @returns {string}
+ */
+exports.getKeyEncoding = function() {
+    return keyEncoding;
+};
+
+/**
+ * setEncoding
+ * Helper to set the encoding being used for keys
+ * @param enc
+ */
+exports.setKeyEncoding = function(enc) {
+    keyEncoding = Buffer.isEncoding(enc) ? enc : keyEncoding;
+};
+
+/**
+ * createSalt
+ * Helper method that returns a salt
+ * @returns string
+ * @throws error
+ */
+exports.createSalt = function(length) {
+    try {
+        return crypto.randomBytes(length);
+    } catch (ex) {
+        console.error('Problem reading random data and generating salt!');
+        throw ex;
+    }
+};
+
+/**
+ * createKeyBuffer
+ * Method which returns a buffer representing a secure key generated with PBKDF2
+ * @returns Buffer
+ */
+exports.createKeyBuffer = function() {
```

Comment:
The key generated here cannot be retrieved again, because the salt is random. When you restart the application, you would lose the key. Is it possible to use the key as the userId to start with instead of randomly creating it. Because the key will come from the backend in the future. We basicly need to be able to do the encryption when a key and a folder is given. So we do not need the pbkdf2 algorithm at all.